### PR TITLE
[Mips] Fix missing sign extension in expansion of sub-word atomic max

### LINKED ISSUE #61881: [MIPS] Incorrect expansion of sub-word signed atomic max
## IR

```llvm
define void @foo(ptr %a, i32 %x) {
  %val = trunc i32 %x to i8
  %1 = atomicrmw max ptr %a, i8 %val seq_cst
  ret void
}
```

## Asm (`-march=mipsel`)

```asm
foo:                                    # @foo
        sync
        addiu   $1, $zero, -4
        and     $1, $4, $1
        andi    $2, $4, 3
        sll     $2, $2, 3
        ori     $3, $zero, 255
        sllv    $3, $3, $2
        nor     $4, $zero, $3
        sllv    $5, $5, $2
$BB0_1:                                 # =>This Inner Loop Header: Depth=1
        ll      $7, 0($1)
        and     $7, $7, $3 # <== WRONG, zero-extending
        and     $5, $5, $3 # <== WRONG, zero-extending 
        slt     $10, $7, $5
        move    $8, $7
        movn    $8, $5, $10
        and     $8, $8, $3
        and     $9, $7, $4
        or      $9, $9, $8
        sc      $9, 0($1)
        beqz    $9, $BB0_1
        nop
        and     $6, $7, $3
        srlv    $6, $6, $2
        sll     $6, $6, 16
        sra     $6, $6, 16
        sync
        jr      $ra
        nop
```

## The problem

The expansion of `i8` signed atomic max masks out all bits that are not involved, but fails to sign-extend the values before passing them to `slt` for a proper signed comparison.

Downstream bug: https://github.com/rust-lang/rust/issues/100650

### PR DESCRIPTION
**Pull request by yingopq**
Add sign extension "SEB/SEH" before compare.

Fix #61881

### CODE REVIEW

**Comment by yingopq** [comment]
@topperc Could you help review this patch? Thanks.

**Comment by brad0** [comment]
@topperc @jyknight Ping.

**Comment by yingopq** [comment]
@MaskRay Hello, could you help review this patch at your convenience? Thanks.

**Comment by brad0** [comment]
@MaskRay ?

File: llvm/lib/Target/Mips/MipsExpandPseudo.cpp [hunk 1]
```diff
@@ -467,8 +481,34 @@ bool MipsExpandPseudo::expandAtomicBinOpSubword(
...
+            .addReg(ShiftAmnt);
+        BuildMI(loopMBB, DL, TII->get(Mips::SRAV), Incr)
+            .addReg(Incr)
+            .addReg(ShiftAmnt);
+        if (STI->hasMips32r2()) {
+          BuildMI(loopMBB, DL, TII->get(SEOp), OldVal).addReg(OldVal);
+          BuildMI(loopMBB, DL, TII->get(SEOp), Incr).addReg(Incr);
+        } else {
+          const unsigned ShiftImm = SEOp == Mips::SEH ? 16 : 24;
+          BuildMI(loopMBB, DL, TII->get(Mips::SLL), OldVal)
```
Comment (topperc) [inline llvm/lib/Target/Mips/MipsExpandPseudo.cpp:497]: Do any of the tests cover this code? I only saw SEH and SEB.

**Comment by brad0** [comment]
@yingopq Ping.

File: llvm/lib/Target/Mips/MipsExpandPseudo.cpp (hunk 1 above)
Comment (yingopq) [reply llvm/lib/Target/Mips/MipsExpandPseudo.cpp:497]: I would add test `-mcpu=mips` in llvm/test/CodeGen/Mips/atomic-min-max.ll to cover `else` branch code.

**Comment by brad0** [comment]
@topperc @MaskRay

**Comment by brad0** [comment]
@topperc @MaskRay

File: llvm/test/CodeGen/Mips/atomic-min-max.ll [hunk 2]
```diff
@@ -980,6 +1049,55 @@ define i16 @test_max_16(ptr nocapture %ptr, i16 signext %val) {
...
+; MIPS32-NEXT:    andi $1, $4, 3
+; MIPS32-NEXT:    sll $10, $1, 3
+; MIPS32-NEXT:    ori $1, $zero, 65535
+; MIPS32-NEXT:    sllv $8, $1, $10
+; MIPS32-NEXT:    nor $9, $zero, $8
+; MIPS32-NEXT:    sllv $7, $5, $10
+; MIPS32-NEXT:  $BB4_1: # %entry
+; MIPS32-NEXT:    # =>This Inner Loop Header: Depth=1
+; MIPS32-NEXT:    ll $2, 0($6)
+; MIPS32-NEXT:    and $2, $2, $8
```
Comment (topperc) [inline llvm/test/CodeGen/Mips/atomic-min-max.ll:1070]: Are these `and` instructions needed? The srav will move the 16 bit chunk to the LSBs, then the sll+sra pair will fill the upper 16 bits with the sign of the lower 16 bits. Doesn't seem like anything needs ANDed before hand.

**Comment by brad0** [comment]
@topperc

**Review by topperc — APPROVED** [review]
LGTM

**Comment by martn3** [comment]
Hello, I started to see MIPS failures in the Rust standard library tests after this change. For example, this program begins to fail after this change:

```rs
fn main() {
    let x = std::sync::atomic::AtomicI8::new(23);
    assert_eq!(x.fetch_max(42, std::sync::atomic::Ordering::SeqCst), 23);
}
```

Please see https://github.com/rust-lang/rust/issues/123772 for more details.

This seems quite problematic, so please let me know if I can help investigate this somehow.

**Comment by yingopq** [comment]
> Hello, I started to see MIPS failures in the Rust standard library tests after this change. For example, this program begins to fail after this change:
> 
> ```rust
> fn main() {
>     let x = std::sync::atomic::AtomicI8::new(23);
>     assert_eq!(x.fetch_max(42, std::sync::atomic::Ordering::SeqCst), 23);
> }
> ```
> 
> Please see [rust-lang/rust#123772](https://github.com/rust-lang/rust/issues/123772) for more details.
> 
> This seems quite problematic, so please let me know if I can help investigate this somehow.

hi, please use code https://github.com/llvm/llvm-project/commit/0e501dbd932ef1c6f4e747c83bf33beef0a09ecf to test again, because we submit another patch after https://github.com/llvm/llvm-project/commit/fbb27d16fa12aa595cbd20a1fb5f1c5b80748fa4. This issue include two patch.
Thanks, please let me know if you have any other questions.

**Comment by nikic** [comment]
@yingopq As I understand it, this was reproduced with the release/18.x branch (or rust's fork, which is essentially the same), which already contains that second commit as well.

**Comment by yingopq** [comment]
> @yingopq As I understand it, this was reproduced with the release/18.x branch (or rust's fork, which is essentially the same), which already contains that second commit as well.

In https://github.com/rust-lang/rust/issues/123772, he comments ```With src/llvm-project pointing to https://github.com/llvm/llvm-project/commit/fbb27d16fa12aa595cbd20a1fb5f1c5b80748fa4 the test also fails, but when pointing src/llvm-project to the parent commit (https://github.com/llvm/llvm-project/commit/e74c1678231ab2a7c6defe8bb90366e664780d6f) the test passes.```. So I reply him use code 0e501dbd932ef1c6f4e747c83bf33beef0a09ecf

**Comment by yingopq** [comment]
> Hi @yingopq, thank you for the reply and for the tip. I double-checked, and I get the test failure also with [0e501db](https://github.com/llvm/llvm-project/commit/0e501dbd932ef1c6f4e747c83bf33beef0a09ecf). I also tested some extra commits just to be sure. Here are my results:
> 
> Commit	Commit info	Commit Title	Test result
> [dc39028](https://github.com/llvm/llvm-project/commit/dc39028906ba4196c3ba544c43ef6b428cf47c51)	branch main	[mlir] Fix -Wsign-compare in ValueBoundsOpInterface.cpp (NFC)	Fail
> [b6ebea7](https://github.com/llvm/llvm-project/commit/b6ebea7972cd05a8e4dcf0d5a54f2c793999995a)	branch release/18.x	[SPARC] Implement L and H inline asm argument modifiers	Fail
> [7c7c3d4](https://github.com/llvm/llvm-project/commit/7c7c3d4a3a278f70b82b4edf5bad8915d929c97d)	tag llvmorg-18.1.2 + 13	[ODS][NFC] Cast range.size() to int32_t in accumulation	Fail
> [0e501db](https://github.com/llvm/llvm-project/commit/0e501dbd932ef1c6f4e747c83bf33beef0a09ecf)	tag llvmorg-18.1.2 + 12	[Mips] Restore wrong deletion of instruction 'and' in unsigned min/max processing.	Fail
> [fbb27d1](https://github.com/llvm/llvm-project/commit/fbb27d16fa12aa595cbd20a1fb5f1c5b80748fa4)	tag llvmorg-18.1.2 + 11	[Mips] Fix missing sign extension in expansion of sub-word atomic max	Fail
> [e74c167](https://github.com/llvm/llvm-project/commit/e74c1678231ab2a7c6defe8bb90366e664780d6f)	tag llvmorg-18.1.2 + 10	[PowerPC] provide CFI for ELF32 to unwind cr2, cr3, cr4	OK
> [1c7c16e](https://github.com/llvm/llvm-project/commit/1c7c16e58240b649254536317160109478c156d1)	tag llvmorg-18.1.2 + 9	[NFC][PowerPC] use script to regenerate the CHECK lines	OK
> Please let me know if you would like me to perform more tests.

Thanks, I would check it.

**Comment by DianQK** [comment]
I have reduced to the following code:

```llvm
target datalayout = "e-m:m-p:32:32-i8:8:32-i16:16:32-i64:64-n32-S64"
target triple = "mipsel-unknown-linux-gnu"

define i32 @main() {
  %1 = call signext i8 @test()
  %2 = sext i8 %1 to i32
  ret i32 %2
}

define signext i8 @test() {
  %i = alloca i8, align 1
  store i8 23, ptr %i, align 1
  %i1 = atomicrmw max ptr %i, i8 42 seq_cst, align 1
  ret i8 %i1
}
```

@yingopq I hope this can be of some use.

**Comment by DianQK** [comment]
Clarify: I don't have any knowledge about MIPS, and the following is only confirmed from the output logs. Please correct me if there is any mistake. :)

TL;DR: All issues have been fixed in LLVM 17. This appears to be an incorrect patch; we might consider reverting it. But I hope @yingopq and others can come to confirm this.

Consider the following two pieces of code:

```llvm
; v1.ll https://github.com/rust-lang/rust/issues/100650
define i8 @test(ptr %arg) {
  %i6 = atomicrmw max ptr %arg, i8 -25 seq_cst, align 1
  %i7 = load atomic i8, ptr %arg seq_cst, align 1
  ret i8 %i7
}

define i32 @main() {
  %i = alloca i8, align 1
  store i8 30, ptr %i, align 1
  %1 = call i8 @test(ptr %i)
  %2 = sext i8 %1 to i32
  ret i32 %2
}
```

and

```llvm
; v2.ll https://github.com/rust-lang/rust/issues/123772
define i8 @test(ptr %arg) {
  %i6 = atomicrmw max ptr %arg, i8 42 seq_cst, align 1
  ret i8 %i6
}

define i32 @main() {
  %i = alloca i8, align 1
  store i8 23, ptr %i, align 1
  %1 = call signext i8 @test(ptr %i)
  %2 = sext i8 %1 to i32
  ret i32 %2
}
```

I can confirm that the return value of `v2.ll` from llvmorg-17-init to 18.1.2 is always 23. However, starting from 18.1.3, it changes to 0.
For `v1.ll`, starting with the commit c65b4d64d4b09795fe237b62a4226121c5b13248, the return value changes from 231 to 30. However, starting from 18.1.3, it also changes to 0.

This is my verification script:

```bash
LLC=path/llc
MIPSEL_GCC=path/mipsel-unknown-linux-gnu-gcc

$LLC -march=mipsel -O0 -mcpu=mips32 -verify-machineinstrs v1.ll -filetype=obj -o v1.o
$MIPSEL_GCC -o v1 v1.o
qemu-mipsel v1
mipsel=$?
$LLC -O0 -march=x86-64 v1.ll -filetype=obj -o v1.o
clang -o v1 v1.o
./v1
x86=$?
echo "v1: $mipsel == $x86"

$LLC -march=mipsel -O0 -mcpu=mips32 -verify-machineinstrs v2.ll -filetype=obj -o v2.o
$MIPSEL_GCC -o v2 v2.o
qemu-mipsel v2
mipsel=$?
$LLC -O0 -march=x86-64 v2.ll -filetype=obj -o v2.o
clang -o v2 v2.o
./v2
x86=$?
if [ "$mipsel" != "$x86" ]; then
  bad=$((bad + 1))
fi
echo "v2: $mipsel == $x86"
```

**Comment by yingopq** [comment]
> Clarify: I don't have any knowledge about MIPS, and the following is only confirmed from the output logs. Please correct me if there is any mistake. :)
> 
> TL;DR: All issues have been fixed in LLVM 17. This appears to be an incorrect patch; we might consider reverting it. But I hope @yingopq and others can come to confirm this.
> 
> Consider the following two pieces of code:
> 
> ```llvm
> ; v1.ll https://github.com/rust-lang/rust/issues/100650
> define i8 @test(ptr %arg) {
>   %i6 = atomicrmw max ptr %arg, i8 -25 seq_cst, align 1
>   %i7 = load atomic i8, ptr %arg seq_cst, align 1
>   ret i8 %i7
> }
> 
> define i32 @main() {
>   %i = alloca i8, align 1
>   store i8 30, ptr %i, align 1
>   %1 = call i8 @test(ptr %i)
>   %2 = sext i8 %1 to i32
>   ret i32 %2
> }
> ```
> 
> and
> 
> ```llvm
> ; v2.ll https://github.com/rust-lang/rust/issues/123772
> define i8 @test(ptr %arg) {
>   %i6 = atomicrmw max ptr %arg, i8 42 seq_cst, align 1
>   ret i8 %i6
> }
> 
> define i32 @main() {
>   %i = alloca i8, align 1
>   store i8 23, ptr %i, align 1
>   %1 = call signext i8 @test(ptr %i)
>   %2 = sext i8 %1 to i32
>   ret i32 %2
> }
> ```
> 
> I can confirm that the return value of `v2.ll` from llvmorg-17-init to 18.1.2 is always 23. However, starting from 18.1.3, it changes to 0. For `v1.ll`, starting with the commit [c65b4d6](https://github.com/llvm/llvm-project/commit/c65b4d64d4b09795fe237b62a4226121c5b13248), the return value changes from 231 to 30. However, starting from 18.1.3, it also changes to 0.
> 
> This is my verification script:
> 
> ```shell
> LLC=path/llc
> MIPSEL_GCC=path/mipsel-unknown-linux-gnu-gcc
> 
> $LLC -march=mipsel -O0 -mcpu=mips32 -verify-machineinstrs v1.ll -filetype=obj -o v1.o
> $MIPSEL_GCC -o v1 v1.o
> qemu-mipsel v1
> mipsel=$?
> $LLC -O0 -march=x86-64 v1.ll -filetype=obj -o v1.o
> clang -o v1 v1.o
> ./v1
> x86=$?
> echo "v1: $mipsel == $x86"
> 
> $LLC -march=mipsel -O0 -mcpu=mips32 -verify-machineinstrs v2.ll -filetype=obj -o v2.o
> $MIPSEL_GCC -o v2 v2.o
> qemu-mipsel v2
> mipsel=$?
> $LLC -O0 -march=x86-64 v2.ll -filetype=obj -o v2.o
> clang -o v2 v2.o
> ./v2
> x86=$?
> if [ "$mipsel" != "$x86" ]; then
>   bad=$((bad + 1))
> fi
> echo "v2: $mipsel == $x86"
> ```

Thanks for your detail comment.
The reason about submitting this patch is that code did not do correct sign extension.  And when I verify this patch, I use ``` %i = alloca i32, align 4 ```, so did not find the current issue you found.  And I am investigating.

**Comment by DianQK** [comment]
Thank you very much for looking into it. Would you mind temporarily reverting these two commits? :>

**Comment by topperc** [comment]
There's definitely an issue in this sequence

```
; MIPSEL-NEXT:    srav $2, $2, $10
; MIPSEL-NEXT:    srav $7, $7, $10
; MIPSEL-NEXT:    seh $2, $2
; MIPSEL-NEXT:    seh $7, $7
; MIPSEL-NEXT:    slt $5, $2, $7
; MIPSEL-NEXT:    move $3, $2
; MIPSEL-NEXT:    movn $3, $7, $5
```

The move and movn need to use the values in $2 and $7 before the srav and seh. The rest of the code expect the 16 bit value to be in the same place in the register as it was when it came out of the load.

**Comment by yingopq** [comment]
> There's definitely an issue in this sequence
> 
> ```
> ; MIPSEL-NEXT:    srav $2, $2, $10
> ; MIPSEL-NEXT:    srav $7, $7, $10
> ; MIPSEL-NEXT:    seh $2, $2
> ; MIPSEL-NEXT:    seh $7, $7
> ; MIPSEL-NEXT:    slt $5, $2, $7
> ; MIPSEL-NEXT:    move $3, $2
> ; MIPSEL-NEXT:    movn $3, $7, $5
> ```
> 
> The move and movn need to use the values in $2 and $7 before the srav and seh. The rest of the code expect the 16 bit value to be in the same place in the register as it was when it came out of the load.

@topperc Thanks so much for pointing out the key to the problem, and I have verified that this is indeed the problem. I would submit patch again.
